PowerShellTools/Intellisense/IntelliSenseManager.cs

```diff
@@ -79,60 +79,84 @@ public int Exec(ref Guid pguidCmdGroup, uint nCmdId, uint nCmdexecopt, IntPtr pv
                     {
-                        Log.Debug("Commit");
-                        _activeSession.Commit();
-
-                        bool isCompletionFullyMatched = false;
-                        _textView.TextBuffer.Properties.TryGetProperty(BufferProperties.SessionCompletionFullyMatchedStatus, out isCompletionFullyMatched);
+                        //if the selection is fully selected, commit the current session 
+                        if (_activeSession.SelectedCompletionSet.SelectionStatus.IsSelected)
+                        {
+                            Log.Debug("Commit");
+                            _activeSession.Commit();
 
-                        if (isCompletionFullyMatched && char.IsWhiteSpace(typedChar))
+                            //also, don't add the character to the buffer 
+                            return VSConstants.S_OK;
+                        }
+                        else
                         {
-                            // If user types all characters in a completion and click Space, then we should commit the selection and add the Space into text buffer
-                            return NextCommandHandler.Exec(ref pguidCmdGroup, nCmdId, nCmdexecopt, pvaIn, pvaOut);
+                            Log.Debug("Dismiss");
+                            //if there is no selection, dismiss the session
+                            _activeSession.Dismiss();
                         }
-
-                        //also, don't add the character to the buffer 
+                    }
+                    else if (nCmdId == (uint)VSConstants.VSStd2KCmdID.TAB && _isRepl)
+                    {
+                        TriggerCompletion();
                         return VSConstants.S_OK;
                     }
+                    break;
+
+                case (uint)VSConstants.VSStd2KCmdID.COMPLETEWORD:
+                    if (_activeSession != null && !_activeSession.IsDismissed)
+                    {
+                        if (_activeSession.CompletionSets[0].SelectionStatus.IsSelected)
+                        {
+                            _activeSession.Commit();
+                        }
+                    }
                     else
                     {
-                    Log.Debug("Dismiss");
-                    //if there is no selection, dismiss the session
-                    _activeSession.Dismiss();
-                }
-                }
-                else if (nCmdId == (uint)VSConstants.VSStd2KCmdID.TAB && _isRepl)
-                {
-                    TriggerCompletion();
+                        TriggerCompletion();
+                    }
                     return VSConstants.S_OK;
-                }
+                default:
+                    break;
+                    
             }
-            // check for Ctrl-Space usage
-            if (commandId == (uint)VSConstants.VSStd2KCmdID.COMPLETEWORD)

```

Comment:
Sure, I refactored this to switch-case. With so many conditions check, I prefer switch, which is much cleaner.